### PR TITLE
Build command does incremental compilation

### PR DESCRIPTION
**Pull request by cdmihai**
Introduce incremental compilation for Build that passes the basic E2E helloworld tests:

- Build and Compile share the code that gets them input or output paths
- Build uses the shared code to check if there are any inputs that are newer than the oldest output
- Added some end to end tests that assert a few incremental scenarios
- Introduced a new --force-incremental-unsafe flag for Build, if users wish to mark a build as unsafe (and thus prevent incremental compilation)

CLI itself still does not self host with incrementality because the postcompile `place-binary` script prevents it. A future PR will address CLI incremental selfhost.

### CODE REVIEW

File: scripts/test/runtests.ps1 [hunk 1]
```diff
@@ -44,7 +49,7 @@ pushd "$TestBinRoot"
 
 # Run each test project
 $TestProjects | ForEach-Object {
-    & "corerun.exe"  "xunit.console.netcore.exe" "$_.dll" -xml "$_.xml" -notrait category=failing
+    & "corerun.exe"  "xunit.console.netcore.exe" "$_.dll" -xml "$_.xml" -notrait category=failing -parallel none
```
Comment (davidfowl) [inline scripts/test/runtests.ps1]: Why are you turning off parallel tests?

File: src/Microsoft.DotNet.Tools.Builder/CompileContext.cs [hunk 2]
```diff
@@ -23,25 +26,25 @@ internal class CompileContext
...
 
-        public bool IsSafeForIncrementalCompilation => _preconditions.PreconditionsDetected();
+        public bool IsSafeForIncrementalCompilation => !_preconditions.PreconditionsDetected();
 
         public CompileContext(ProjectContext rootProject, BuilderCommandApp args)
         {
             _rootProject = rootProject;
-            _args = args;
+
+            //Cleaner to clone the args and mutate the clone than have separate CompileContext fields for mutated args
```
Comment (davidfowl) [inline src/Microsoft.DotNet.Tools.Builder/CompileContext.cs]: nit: Put a space between `// and the comment`

File: src/Microsoft.DotNet.Tools.Compiler/CompilerUtil.cs [hunk 3]
```diff
@@ -113,6 +115,46 @@ public static string GetResourceFileMetadataName(Project project, KeyValuePair<s
             return name;
         }
 
+        //used in incremental compilation
+        public static IEnumerable<string> GetCompilationSources(ProjectContext project) => project.ProjectFile.Files.SourceFiles;
+
+
```
Comment (davidfowl) [inline src/Microsoft.DotNet.Tools.Compiler/CompilerUtil.cs]: What's with all the extra space?

File: src/Microsoft.DotNet.Tools.Compiler/CompilerUtil.cs [hunk 4]
```diff
@@ -113,6 +115,46 @@ public static string GetResourceFileMetadataName(Project project, KeyValuePair<s
...
+
+        //used in incremental compilation
+        public static string GetCompilationOutput(Project project, NuGetFramework framework, string configuration, string outputPath)
```
Comment (davidfowl) [inline src/Microsoft.DotNet.Tools.Compiler/CompilerUtil.cs]: Doesn't this helper exist already? /cc @piotrpMSFT

File: test/E2E/E2ETest.cs [hunk 5]
```diff
@@ -50,6 +51,76 @@ public void TestDotnetCompile()
             TestOutputExecutable(OutputDirectory);
         }
 
+        [Fact] public void TestDotnetIncrementalBuild()
```
Comment (davidfowl) [inline test/E2E/E2ETest.cs]: Fix the formatting. Put Fact on top of the method like the rest of the tests.

File: src/Microsoft.DotNet.Tools.Compiler/CompilerUtil.cs [hunk 6]
```diff
@@ -2,6 +2,7 @@
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
```
Comment (davidfowl) [inline src/Microsoft.DotNet.Tools.Compiler/CompilerUtil.cs]: This class is beginning to look like a dumping ground for helper methods...

File: src/Microsoft.DotNet.Tools.Builder/ProjectDependenciesFacade.cs [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+namespace Microsoft.DotNet.Tools.Build
+{
+    //facade over the dependencies of a project context
+    internal class ProjectDependenciesFacade
+    {
+        //projectName -> ProjectDescription
+        public Dictionary<string, ProjectDescription> ProjectDependenciesWithSources { get; }
+        public List<LibraryExport> Dependencies { get; }
+
+        public ProjectDependenciesFacade(ProjectContext _rootProject, string configValue)
```
Comment (davidfowl) [inline src/Microsoft.DotNet.Tools.Builder/ProjectDependenciesFacade.cs]: Why is there an _ in the parameter name?

File: src/Microsoft.DotNet.Tools.Builder/CompileContext.cs [hunk 8]
```diff
@@ -52,22 +55,85 @@ public bool Compile(bool incremental)
...
             PrintSummary(success);
 
             return success;
         }
 
+        private bool NeedsRebuilding(ProjectContext project, ProjectDependenciesFacade dependencies)
+        {
+            var compilerIO = GetCompileIO(project, _args.ConfigValue, _args.OutputValue, _args.IntermediateValue, dependencies);
+
+            if (! (compilerIO.Outputs.Any() && compilerIO.Inputs.Any()))
```
Comment (davidfowl) [inline src/Microsoft.DotNet.Tools.Builder/CompileContext.cs]: Couldn't you do all of this logic with a single for loop over inputs and outputs?

File: src/Microsoft.DotNet.Tools.Builder/CompileContext.cs [hunk 9]
```diff
@@ -52,22 +55,85 @@ public bool Compile(bool incremental)
             CreateOutputDirectories();
 
             //compile dependencies
-            foreach (var dependency in Sort(_dependencies))
+            foreach (var dependency in Sort(_dependencies.ProjectDependenciesWithSources))
             {
-                if (!InvokeCompileOnDependency(dependency, _outputPath, _intermediateOutputPath))
+                if (incremental)
+                {
+                    var dependencyContext = ProjectContext.Create(dependency.Path, dependency.TargetFrameworkInfo.FrameworkName);
```
Comment (davidfowl) [inline src/Microsoft.DotNet.Tools.Builder/CompileContext.cs]: `dependency.TargetFrameworkInfo.FrameworkName` isn't correct. It should be using `dependency.Framework`

File: src/Microsoft.DotNet.Tools.Builder/CompileContext.cs [hunk 10]
```diff
@@ -52,22 +55,85 @@ public bool Compile(bool incremental)
...
+        {
+            var compilerIO = GetCompileIO(project, _args.ConfigValue, _args.OutputValue, _args.IntermediateValue, dependencies);
+
+            if (! (compilerIO.Outputs.Any() && compilerIO.Inputs.Any()))
+            {
+                Reporter.Verbose.WriteLine($"\nProject {project.ProjectName()} was compiled because it either has empty inputs or outputs");
+                return true;
+            }
+
+            if (compilerIO.Outputs.Any(o => ! File.Exists(o)))
```
Comment (davidfowl) [inline src/Microsoft.DotNet.Tools.Builder/CompileContext.cs]: nit; The space after `! File.Exists()` should be removed (all over the code base)

File: src/Microsoft.DotNet.Tools.Builder/CompileContext.cs [hunk 11]
```diff
@@ -248,6 +293,103 @@ private static void Sort(ProjectDescription project, Dictionary<string, ProjectD
...
+
+            Debug.Assert(compilerIO.Inputs.All(p => !Directory.Exists(p)), "Directories detected among compilation IO paths");
+            Debug.Assert(compilerIO.Outputs.All(p => !Directory.Exists(p)), "Directories detected among compilation IO paths");
+
+            foreach (var input in compilerIO.Inputs)
+            {
+                Debug.Assert(File.Exists(input), $"Input {input} in project {project.ProjectName()} does not exist");
+            }
+
+            return compilerIO;
```
Comment (cdmihai) [inline src/Microsoft.DotNet.Tools.Builder/CompileContext.cs]: The above assertions need to be moved into NeedsRebuilding and force a rebuild.

File: scripts/test/runtests.ps1 (hunk 1 above)
Comment (cdmihai) [inline scripts/test/runtests.ps1]: I was having some issues with tests so I preventively turned it off. However it turned out my tests were failing because the stage 0 was used and not stage 2 :)

I left them turned off as a safety measure in case tests from different test classes might read / write on the same paths which would lead to races and weird test failures.
Until CLI tests use a shared mechanism that guarantees the absence file IO races, I think it's safer to not have parallel tests.

File: src/Microsoft.DotNet.Tools.Compiler/CompilerUtil.cs (hunk 3 above)
Comment (cdmihai) [inline src/Microsoft.DotNet.Tools.Compiler/CompilerUtil.cs]: Probably a remnant from moving a lot of code around. Thanks for pointing that out

File: src/Microsoft.DotNet.Tools.Compiler/CompilerUtil.cs (hunk 4 above)
Comment (cdmihai) [inline src/Microsoft.DotNet.Tools.Compiler/CompilerUtil.cs]: I moved it here from Compile.

File: test/E2E/E2ETest.cs (hunk 5 above)
Comment (cdmihai) [inline test/E2E/E2ETest.cs]: Good catch. Thanks.

File: src/Microsoft.DotNet.Tools.Builder/ProjectDependenciesFacade.cs (hunk 7 above)
Comment (cdmihai) [inline src/Microsoft.DotNet.Tools.Builder/ProjectDependenciesFacade.cs]: I rewrote the way that class got instantiated a few times. A field name must have slipped in. Thanks for pointing that out.

File: src/Microsoft.DotNet.Tools.Builder/CompileContext.cs (hunk 8 above)
Comment (cdmihai) [inline src/Microsoft.DotNet.Tools.Builder/CompileContext.cs]: True. However, right now I am still fleshing out the desired behaviour on edge cases and it helps to have them separated. I'd leave this for future refactorings when specs are more clear.

File: src/Microsoft.DotNet.Tools.Builder/CompileContext.cs (hunk 9 above)
Comment (cdmihai) [inline src/Microsoft.DotNet.Tools.Builder/CompileContext.cs]: OK. I wasn't sure what's the right way to get it. Is there a difference between the two? Is it OK if I replace it everywhere I get the framework out of a dependency?

Comment (davidfowl) [inline src/Microsoft.DotNet.Tools.Builder/CompileContext.cs]: That's the right way.

File: src/Microsoft.DotNet.Tools.Builder/CompileContext.cs [hunk 12]
```diff
@@ -52,22 +55,85 @@ public bool Compile(bool incremental)
...
+                Reporter.Verbose.WriteLine($"\nProject {project.ProjectName()} was compiled because expected outputs were missing");
+                return true;
+            }
+
+            var minOutputPath = compilerIO.Outputs.First();
+            var minDate = File.GetLastWriteTime(minOutputPath);
+
+            foreach (var outputPath in compilerIO.Outputs)
+            {
+                if (File.GetLastWriteTime(outputPath) >= minDate) continue;
```
Comment (davidfowl) [inline src/Microsoft.DotNet.Tools.Builder/CompileContext.cs]: No self closing if statements. Just add braces

File: src/Microsoft.DotNet.Tools.Compiler/CompilerUtil.cs (hunk 6 above)
Comment (cdmihai) [inline src/Microsoft.DotNet.Tools.Compiler/CompilerUtil.cs]: I am using this class to hold the functionality that needs to be available to both build and compile. Currently I'm using it to learn and collect what the common interface is. Ideally, this communication should be done via something more first class than a utility class (an interface maybe?).

For now, I'd rather keep it like this until we get a better idea on how the 2 verbs need to communicate.

**Comment by cdmihai** [comment]
@dotnet-bot test this please

**Comment by cdmihai** [comment]
Addressed David's comments (except the ones that still show up in the PR; They'll go in future work once incremental story is more mature)

**Comment by davidfowl** [comment]
@cdmihai what;s the deal with this?

**Comment by cdmihai** [comment]
@davidfowl 
This PR introduces incremental compilation that passes the basic E2E helloworld tests. At the end of this week I'll post another PR that enables the entire CLI to build itself incrementally.

Regarding your prior feedback that I did not fix yet:

- NeedsRebuild needs some refactoring, true; However, for now, I am working on making sure CLI selfhosts incrementaly (this week's milestone) and then, during testing and stabilization I want to refactor things (fix as many of the todos in the code as possible)

Concerning CompilerUtil, it's purpose is to contain the functionality that's shared between build and compile (build needs to call it to compute incremental checks and make decisions on when to invoke compile). It consists of code I moved from inside Compile to CompilerUtil. Ideally, this would be an interface that both Compile and Build use. Since all the utility methods receive a ProjectContext, the utility methods could also compile / build specific extension methods on ProjectContext. Even more ideally, if we want to make the entire CLI incremental, there would be some higher level interface that all verbs need to implement. For now, I want to keep CompilerUtil as it is to document the shared functionality and decide later (during testing and stabilization) on how to best refactor it (maybe I shouold rename it to something else to make it clear that it only contains shared stuff).

File: scripts/test/runtests.ps1 (hunk 1 above)
Comment (davidfowl) [inline scripts/test/runtests.ps1]: If our tests are written poorly I agree it'll happen. I'd revert this change as its unrelated to your changes.

Comment (davidfowl) [inline scripts/test/runtests.ps1]: He stage 0 stage 2 thing is a bug and should be fixed

File: scripts/test/runtests.sh [hunk 13]
```diff
@@ -6,6 +6,12 @@
 
 set -e
 
+#force tests to use stage 2 output
+export StartPath=$PATH
```
Comment (brthor) [inline scripts/test/runtests.sh]: This file should be agnostic to which version of the binaries it is testing. 

It will be used to test binaries in packages and other places than the stage2 directory. 
This logic should be in the calling script.

File: scripts/test/runtests.ps1 [hunk 14]
```diff
@@ -5,6 +5,11 @@
 
 . "$PSScriptRoot\..\common\_common.ps1"
 
+#force tests to use stage 2 output
+$StartPath = $env:PATH
```
Comment (brthor) [inline scripts/test/runtests.ps1]: This file should be agnostic to which version of the binaries it is testing.

File: test/E2E/E2ETest.cs [hunk 15]
```diff
@@ -50,6 +51,78 @@ public void TestDotnetCompile()
             TestOutputExecutable(OutputDirectory);
         }
 
+        [Fact] 
+        public void TestDotnetIncrementalBuild()
+        {
+            TestSetup();
+
+            //first build
+            TestRunCommand("dotnet", $" --verbose build --build-profile -o {OutputDirectory}");
```
Comment (brthor) [inline test/E2E/E2ETest.cs]: Please update these tests to use the Test infrastructure similar to the new factoring in E2E

**Comment by cdmihai** [comment]
@brthor @davidfowl 

Regarding running tests for stage2. Should I do a different PR where I solve that issue, since the bug was there before this PR, but is blocking it?

**Comment by brthor** [comment]
@cdmihai What is the bug exactly? 

I'm not against forcing the tests to run against stage2, I just think that the path should be set before calling the `runtests` scripts so that script becomes reusable.

For example, I am hoping to use that script to run all tests against the binaries installed by the debian package as a sort of verification against packaging bugs.

File: test/E2E/E2ETest.cs (hunk 15 above)
Comment (brthor) [inline test/E2E/E2ETest.cs]: This should not be too bad, but I realized the --build-profile flag will need to be added to BuildCommand.cs. All other flags should be there:

https://github.com/dotnet/cli/blob/master/test/Microsoft.DotNet.Tools.Tests.Utilities/Commands/BuildCommand.cs

**Comment by davidfowl** [comment]
We need to merge something as soon as possible because the ASP.NET builds are hella slowing down nexus of the CLi switch. We also need to allow passing multiple projects to build so we can build thr graph in a single call.

File: scripts/test/runtests.ps1 (hunk 1 above)
Comment (piotrpMSFT) [inline scripts/test/runtests.ps1]: Agreed. Let's leave parallel on. If someone writes a buggy test they will fix it.

File: scripts/test/runtests.sh (hunk 13 above)
Comment (piotrpMSFT) [inline scripts/test/runtests.sh]: Agreed

File: scripts/test/runtests.sh [hunk 16]
```diff
@@ -42,7 +48,7 @@ failCount=0
 
 for project in ${TestProjects[@]}
 do
-    ./corerun  "xunit.console.netcore.exe" "$project.dll" -xml "project.xml" -notrait category=failing
+    ./corerun  "xunit.console.netcore.exe" "$project.dll" -xml "project.xml" -notrait category=failing -parallel none
```
Comment (piotrpMSFT) [inline scripts/test/runtests.sh]: Parallel on plz

File: src/Microsoft.DotNet.Tools.Builder/CompileContext.cs [hunk 17]
```diff
@@ -14,6 +15,7 @@
 
 namespace Microsoft.DotNet.Tools.Build
 {
+    // todo: Convert CompileContext into a DAG of dependencies: if a node needs recompilation, the entire path up to root needs compilation
```
Comment (piotrpMSFT) [inline src/Microsoft.DotNet.Tools.Builder/CompileContext.cs]: Issue #?

File: scripts/test/runtests.ps1 (hunk 1 above)
Comment (cdmihai) [inline scripts/test/runtests.ps1]: Alright. I'll leave them turned off.

**Comment by cdmihai** [comment]
@brthor 

On the powershell scripts, stage0 is used to run the tests, not stage2. I'll fixing that (telling runntests.ps1 which stage to test) in a separate PR as suggested.

**Comment by cdmihai** [comment]
Addressed feedback:

- tests are run in parallel
- separated passing the bin dir to test in another PR: #748
- rewrote the incremental tests to use the new testing framework

Currently blocked by #748. Can merge once that PR is in.

**Comment by cdmihai** [comment]
@dotnet-bot test this please

**Comment by cdmihai** [comment]
@dotnet-bot test this please

**Comment by cdmihai** [comment]
Closing this PR. It is subsumed by #761
